pkg/location/type_git.go

```diff
@@ -0,0 +1,123 @@
+package location
+
+import (
+	"fmt"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/GoogleContainerTools/kpt/internal/errors"
+	"github.com/GoogleContainerTools/kpt/internal/util/parse"
+)
+
+type Git struct {
+	// Repo is the git repository the package.
+	// e.g. 'https://github.com/kubernetes/examples.git'
+	Repo string
+
+	// Directory is the sub directory of the git repository.
+	// e.g. 'staging/cockroachdb'
+	Directory string
+
+	// Ref can be a Git branch, tag, or a commit SHA-1.
+	Ref string
+}
+
+var _ Reference = Git{}
+
+type GitLock struct {
+	Git
+
+	// Commit is the SHA-1 for the last fetch of the package.
+	// This is set by kpt for bookkeeping purposes.
+	Commit string
+}
+
+var _ ReferenceLock = GitLock{}
+
+func NewGit(location string, opts ...Option) (Git, error) {
+	return newGit(location, makeOptions(opts...))
+}
+
+func newGit(location string, opt options) (Git, error) {
+	// args[1] is "" for commands that do not require an output path
+	gitTarget, err := parse.GitParseArgs(opt.ctx, []string{location, ""})
+	var zero parse.GitTarget
+	if err != nil || gitTarget == zero {
+		return Git{}, err
+	}
+
+	dir := gitTarget.Directory
+	if strings.HasPrefix(dir, "/") {
+		dir, err = filepath.Rel("/", gitTarget.Directory)
+		if err != nil {
+			return Git{}, err
+		}
+	}
+
+	return Git{
+		Repo:      gitTarget.Repo,
+		Directory: dir,
+		Ref:       gitTarget.Ref,
+	}, nil
+}
+
+func parseGit(location string, opt options) (Reference, error) {
+	git, gitErr := newGit(location, opt)
+	var zero Git
+	if gitErr == nil && git != zero {
+		return git, nil
+	}
+
+	// TODO - figure out which gitErr must be returned, and which simply mean "it's not a git path"
```

Comment:
I had similar question reading the parse protocol (first to recognize path wins) and non-recognition is signaled by returning nil; What if there is parser later in the parser list that will succeed?

Reply:
That is a problem. Right now it's a requirement that parsers are added in the right order. That is, put stdio "-" first because it's most specific and "dir" last because it'll accept most values.

But that is a good example, when either "-" and "dir" can be used as an arg, you've more or less lost the ability to read a dir named "-". Well, unless you use "./-" as the path. TBH, that's also similar to the "go cmd X" problem where X has to start with "./" so it's not parsed as a module package name.

An improvement could be to have a priority order, but that breaks down when multiple custom types (that are not aware of each other) that can be solved by having the caller provide the priority for each.. location.WithGit(1), location.WithOci(2) location.WithDir(0) etc... Which would probably seem awful to use...

Reply:
Okay - I pushed a refactor that does a few things.

* there's now a single WithParsers(...) option that takes the ordered `...parsers`. So the order of options is no longer an issue.

* you can now add StdioParser upfront all of the time, so that list no longer needs to vary. Instead, it is only active when you also have the WithStdin and WithStdout option. e.g. You'll have WithStdin option only on the cli args which are input and accept "-".

* the return value is now a result struct {ref, err, helpText} with code comments for semantics. `ref` and `err` are always final, and all of the helptext is returned if no parser returns ref or err.

So `err` means I **know** this is my format, and it's broken. and `helpText` means this _might_ be my format, and here's text to add to the warning if a later parser doesn't lock onto it unambiguously